mlflow/models/signature.py

```diff
@@ -340,35 +351,90 @@ def _extract_type_hints(f, input_arg_index):
+        parameters = inspect.signature(func).parameters
+    return (
+        # predict(self, context, model_input, ...)
+        "context" in parameters
+        # predict(self, ctx, model_input, ...) ctx can be any parameter name
+        or len([param for param in parameters if param != "params"]) == 2
+    )
+
+
+def _infer_signature_from_type_hints(func, type_hints: _TypeHints, input_example=None):
+    if type_hints.input is None:
         return None
 
     params = None
     params_key = "params"
     if _contains_params(input_example):
         input_example, params = input_example
 
-    input_schema = _infer_schema_from_type_hint(hints.input, input_example) if hints.input else None
+    try:
+        input_schema = _infer_schema_from_type_hint(type_hints.input)
+    except InvalidTypeHintException as e:
+        warnings.warn(e.message, stacklevel=2)
+        return None
+    is_output_type_hint_valid = type_hints.output is not None
+    try:
+        output_schema = (
+            _infer_schema_from_type_hint(type_hints.output) if type_hints.output else None
+        )
+    except InvalidTypeHintException as e:
+        warnings.warn(
+            f"Invalid output type hint, setting output schema to AnyType. Error: {e}", stacklevel=2
+        )
+        is_output_type_hint_valid = False
+        output_schema = Schema([ColSpec(type=AnyType())])
     params_schema = _infer_param_schema(params) if params else None
-    input_arg_name = _get_arg_names(func)[input_arg_index]
     if input_example:
-        inputs = {input_arg_name: input_example}
-        if params and params_key in inspect.signature(func).parameters:
-            inputs[params_key] = params
-        # This is for PythonModel's predict function
-        if input_arg_index == 1:
-            inputs["context"] = None
-        output_example = func(**inputs)
-    else:
-        output_example = None
-    output_schema = (
-        _infer_schema_from_type_hint(hints.output, output_example) if hints.output else None
-    )
+        # TODO: we can remove input example validation here
+        # once we move the validation inside `predict` function
+        if msg := _get_example_validation_result(
+            example=input_example, type_hint=type_hints.input
+        ).error_message:
+            _logger.warning(
+                "Input example is not compatible with the type hint of the `predict` function. "
+                f"Error: {msg}"
+            )
+        else:
+            kwargs = {}
+            if params and params_key in inspect.signature(func).parameters:
+                kwargs[params_key] = params
+            # This is for PythonModel's predict function
+            if _is_context_in_predict_function_signature(func=func):
+                inputs = [None, input_example]
+            else:
+                inputs = [input_example]
+            _logger.info("Running the predict function to infer output based on input example")
```

Comment:
```suggestion
            _logger.info("Running the predict function to generate output based on input example")
```

I think `generate` is more accurate here. We don't do any inference here.